server/src/main/java/org/opensearch/common/time/JavaDateFormatter.java

```diff
@@ -285,29 +290,43 @@ public TemporalAccessor parse(String input) {
      * @throws DateTimeParseException when unable to parse with any parsers
      */
     private TemporalAccessor doParse(String input) {
-        if (parsers.size() > 1) {
-            Object object = null;
-            if (canCacheLastParsedFormatter && lastParsedformatter != null) {
-                ParsePosition pos = new ParsePosition(0);
-                object = lastParsedformatter.toFormat().parseObject(input, pos);
-                if (parsingSucceeded(object, input, pos)) {
-                    return (TemporalAccessor) object;
-                }
+        final ParsePosition pos = new ParsePosition(0);
+        //Copy to local from volatile to avoid error prune in multi-threads,
+        //other threads may update lastParsedformatter
+        final DateTimeFormatter cachedFormatter = lastParsedformatter;
+        Object object = null;
+        if (canCacheLastParsedFormatter && cachedFormatter != null) {
+            object = tryParseObject(cachedFormatter, input, pos);
```

Comment:
Thanks for looking into the issue @Pengxiaolong , the `Format::parseObject` does not document any exceptions but explicitly says that `null` is going to be returned:

>      * @return An <code>Object</code> parsed from the string. In case of
     *         error, returns null.

Reply:
Yes, that's what the document says, I suspect it is a bug in JDK anyway. The exception is from class java.time.format.DateTimeFormatter.ClassicFormat, 
       ```
        @Override
        public Object parseObject(String text, ParsePosition pos) {
            Objects.requireNonNull(text, "text");
            DateTimeParseContext context;
            try {
                context = formatter.parseUnresolved0(text, pos);
            } catch (IndexOutOfBoundsException ex) {
                if (pos.getErrorIndex() < 0) {
                    pos.setErrorIndex(0);
                }
                return null;
            }
            if (context == null) {
                if (pos.getErrorIndex() < 0) {
                    pos.setErrorIndex(0);
                }
                return null;
            }
            try {
                TemporalAccessor resolved = context.toResolved(formatter.resolverStyle, formatter.resolverFields);
                if (parseType == null) {
                    return resolved;
                }
                return resolved.query(parseType);
            } catch (RuntimeException ex) {
                pos.setErrorIndex(0);
                return null;
            }
        }
```

In the code above, it calls formatter.parseUnresolved0(text, pos) to parse the text, then try to resolve the parse result. When it call formatter.parseUnresolved0, it doesn't catch RuntimeException which caused the bug, formatter.parseUnresolved0 may throw DataTimeException. 

If you dig into the stack trace, you will find the exception is from here:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/time/format/DateTimeFormatterBuilder.java#L4126

Reply:
Thanks a lot @Pengxiaolong , I've opened https://bugs.openjdk.org/browse/JDK-8319640 to track this, meanwhile I think the problem is legit and has to be mitigated, thank you.